README.md
# compiler-interrupts

[![crates.io](https://img.shields.io/crates/v/compiler-interrupts.svg)][crates.io]
[![docs.rs](https://docs.rs/compiler-interrupts/badge.svg)][docs.rs]
[![license](https://img.shields.io/crates/l/compiler-interrupts.svg)][license]

`compiler-interrupts` provides Rust API for the Compiler Interrupts library.
Check out the Compiler Interrupts [main repository][compiler-interrupts] for more info.

## Requirements

* [Rust 1.45.0][rust] or later is required.
Due to the usage of [`#[thread_local]`][thread_local] unstable feature,
this package currently requires nightly Rust.

## Getting started

Add this to your `Cargo.toml`.

``` toml
[dependencies]
compiler-interrupts = "1.0"
```

Register the Compiler Interrupts handler in your program.

``` rust
#![feature(thread_local)]

#[thread_local]
#[allow(non_upper_case_globals)]
static mut prev_ic: i64 = 0;

fn interrupt_handler(ic: i64) {
    let interval;
    unsafe {
        // save the last interval
        interval = ic - prev_ic;

        // update the instruction count
        prev_ic = ic;
    }
    if interval < 0 {
        panic!("IR interval was negative")
    }
    println!(
        "CI @ {}: last interval = {} IR",
        std::thread::current().name().expect("invalid thread name"),
        interval
    );
}

fn main() {
    // register the CI handler for 1000 IR and cycles interval
    unsafe {
        compiler_interrupts::register(1000, 1000, interrupt_handler);
    }

    // do something compute-intensive
    for _ in 0..100 {}

    println!("i can add an unsafe block, right?")
}
```

If you have [`cargo-compiler-interrupts`][cargo-compiler-interrupts] installed,
you can now run `cargo build-ci` to start the compilation and integration.
Check out the **[documentation][docs.rs]** for more info about the API.

## Contribution

All issue reports, feature requests, pull requests and GitHub stars are welcomed
and much appreciated.

## Author

Quan Tran ([@quanshousio][quanshousio])

## Acknowledgements

* My advisor [Jakob Eriksson][jakob] for the enormous support for this project.
* [Nilanjana Basu][nilanjana] for implementing the Compiler Interrupts.

## License

`compiler-interrupts` is available under the MIT license.
See the [LICENSE][license] file for more info.

[crates.io]: https://crates.io/crates/compiler-interrupts
[docs.rs]: https://docs.rs/compiler-interrupts
[license]: https://github.com/bitslab/compiler-interrupts-rs/blob/main/LICENSE
[compiler-interrupts]: https://github.com/bitslab/CompilerInterrupts
[rust]: https://www.rust-lang.org/tools/install
[thread_local]: https://github.com/rust-lang/rust/issues/29594
[cargo-compiler-interrupts]: https://github.com/bitslab/cargo-compiler-interrupts
[quanshousio]: https://quanshousio.com
[jakob]: https://www.linkedin.com/in/erikssonjakob
[nilanjana]: https://www.linkedin.com/in/nilanjana-basu-99027959


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. `compiler-interrupts` adheres to [Semantic Versioning](https://semver.org/).

#### 1.x Releases

- `1.0.x` Releases - [1.0.0](#100) | [1.0.1](#101)

---

## [1.0.1](https://github.com/bitslab/compiler-interrupts-rs/releases/tag/1.0.1)

Released on 2021-07-31.

#### Added

- Add a new `profiler` example. This example is only available on x86-64 Linux platforms.

#### Updated

- Move `ci-demo` package to `examples/demo`.
- Fix a missing thread-local delcaration for `prev_ic` variable in the `demo` example.

## [1.0.0](https://github.com/bitslab/compiler-interrupts-rs/releases/tag/1.0.0)

Released on 2021-07-24.

#### Added

- Initial release of `compiler-interrupts`.


examples/demo.rs
#![feature(thread_local)]

use anyhow::{Context, Result};
use nanorand::{Rng, WyRand};

const BASE_VAL: i32 = 10000;
const CI_INTERVAL: i64 = 1_000_000;
const MAX_THREADS: i32 = 8;

#[thread_local]
#[allow(non_upper_case_globals)]
static mut prev_ic: i64 = 0;

fn rand() -> i32 {
    let mut rng = WyRand::new();
    rng.generate_range(0..i32::MAX)
}

fn interrupt_handler(ic: i64) {
    let interval;
    unsafe {
        interval = ic - prev_ic;
        prev_ic = ic;
    }
    if interval < 0 {
        panic!("IR interval was negative")
    }
    println!(
        "CI @ {}: last interval = {} IR",
        std::thread::current().name().expect("invalid thread name"),
        interval
    );
}

fn enable_hook() {
    println!("This function is called after CI is enabled");
}

fn disable_hook() {
    println!("This function is called before CI is disabled");
}

fn increment() -> Result<()> {
    unsafe {
        println!("interval: {}", CI_INTERVAL);
        compiler_interrupts::register(CI_INTERVAL, CI_INTERVAL, interrupt_handler);
    }

    let mut counter = 0;
    let iterations = BASE_VAL + (rand() % 10);
    for _ in 0..iterations {
        counter += rand() % 10;
    }
    println!(
        "increment(): thread: {} -> counter: {}",
        std::thread::current()
            .name()
            .context("failed to get thread name")?,
        counter
    );

    Ok(())
}

fn decrement() -> Result<()> {
    unsafe {
        println!("interval: {}", CI_INTERVAL);
        compiler_interrupts::register(CI_INTERVAL, CI_INTERVAL, interrupt_handler);
    }

    // register the enable and disable hooks
    unsafe {
        // `enable_hook` will be called when enable the CI
        compiler_interrupts::register_enable_hook(enable_hook);

        // `disable_hook` will be called when disable the CI
        compiler_interrupts::register_disable_hook(disable_hook);
    }

    // temporarily disable CI for the remaining part of the execution
    unsafe {
        compiler_interrupts::disable();
    }

    let mut counter = 0;
    let iterations = BASE_VAL + (rand() % 10);

    unsafe {
        // do nothing since CI is already disabled
        compiler_interrupts::disable();

        // as there were two disable calls before
        // we have to call enable twice to re-enable CI
        compiler_interrupts::enable();
        compiler_interrupts::enable();
    }

    for _ in 0..iterations {
        counter -= rand() % 10;
    }

    // de-register CI, which effectively disable CI for
    // the remaining part of the execution
    unsafe {
        compiler_interrupts::deregister();
    }

    println!(
        "decrement(): thread: {} -> counter: {}",
        std::thread::current()
            .name()
            .context("failed to get thread name")?,
        counter
    );

    Ok(())
}

fn main() -> Result<()> {
    // register the CI handler
    unsafe {
        compiler_interrupts::register(CI_INTERVAL, CI_INTERVAL, interrupt_handler);
    }

    // check argument for number of threads
    let args = std::env::args().collect::<Vec<_>>();
    let num_threads = if args.len() == 2 {
        let n = args[1].parse().unwrap_or(MAX_THREADS);
        if n > MAX_THREADS {
            println!("max threads: {}", MAX_THREADS);
            MAX_THREADS
        } else {
            n
        }
    } else {
        MAX_THREADS
    };

    // increment threads
    println!("starting {} increment threads", num_threads);
    let mut threads = vec![];
    for thread_id in 0..num_threads {
        let thread = std::thread::Builder::new()
            .name(format!("inc{}", thread_id))
            .spawn(|| increment())
            .context("failed to create thread")?;
        threads.push(thread);
    }
    for thread in threads {
        thread.join().expect("thread panicked")?;
    }

    // decrement threads
    println!("starting {} decrement threads", num_threads);
    let mut threads = vec![];
    for thread_id in 0..num_threads {
        let thread = std::thread::Builder::new()
            .name(format!("dec{}", thread_id))
            .spawn(|| decrement())
            .context("failed to create thread")?;
        threads.push(thread);
    }
    for thread in threads {
        thread.join().expect("thread panicked")?;
    }

    Ok(())
}


ci-demo/src/main.rs
const BASE_VAL: i32 = 10000;
const MAX_THREADS: i32 = 64;

static mut PREVIOUS_IC: i64 = 0;

fn rand() -> i32 {
    use nanorand::{Rng, WyRand};
    let mut rng = WyRand::new();
    rng.generate_range(0..i32::MAX)
}

fn interrupt_handler(ic: i64) {
    unsafe {
        println!(
            "CI @ {}: last interval = {} IR",
            std::thread::current().name().unwrap(),
            ic - PREVIOUS_IC
        );
        PREVIOUS_IC = ic;
    }
}

fn enable_hook() {
    println!("This function is called after CI is enabled");
}

fn disable_hook() {
    println!("This function is called before CI is disabled");
}

fn increment(thread_id: i32) {
    // register the CI handler for 10000 IR and cycles interval
    unsafe {
        compiler_interrupts::register(10000, 10000, interrupt_handler);
    }

    let mut counter = 0;
    let iterations = BASE_VAL + (rand() % 10);
    for _ in 0..iterations {
        counter += rand() % 10;
    }
    println!("increment(): thread: {} -> counter: {}", thread_id, counter);
}

fn decrement(thread_id: i32) {
    // register the CI handler for 10000 IR and cycles interval
    unsafe {
        compiler_interrupts::register(10000, 10000, interrupt_handler);
    }

    // register the enable and disable hooks
    unsafe {
        // `enable_hook` will be called when enable the CI
        compiler_interrupts::register_enable_hook(enable_hook);

        // `disable_hook` will be called when disable the CI
        compiler_interrupts::register_disable_hook(disable_hook);
    }

    // temporarily disable CI for the remaining part of the execution
    unsafe {
        compiler_interrupts::disable();
    }

    let mut counter = 0;
    let iterations = BASE_VAL + (rand() % 10);

    unsafe {
        // do nothing since CI is already disabled
        compiler_interrupts::disable();

        // as there were two disable calls before
        // we have to call enable twice to re-enable CI
        compiler_interrupts::enable();
        compiler_interrupts::enable();
    }

    // de-register CI, which effectively disable CI for
    // the remaining part of the execution
    unsafe {
        compiler_interrupts::deregister();
    }

    for _ in 0..iterations {
        counter -= rand() % 10;
    }
    println!("decrement(): thread: {} -> counter: {}", thread_id, counter);
}

fn main() {
    // register the CI handler for 1000 IR and cycles interval
    unsafe {
        compiler_interrupts::register(1000, 1000, interrupt_handler);
    }

    // check for number of threads argument
    let args = std::env::args().collect::<Vec<_>>();
    let num_threads = if args.len() == 2 {
        let n = args[1].parse().unwrap_or(MAX_THREADS);
        if n > MAX_THREADS {
            MAX_THREADS
        } else {
            n
        }
    } else {
        MAX_THREADS
    };

    // increment threads
    println!("starting {} increment threads", num_threads);
    let mut threads = vec![];
    for thread_id in 0..num_threads {
        let thread = std::thread::Builder::new()
            .name(format!("inc{}", thread_id))
            .spawn(move || increment(thread_id))
            .unwrap();
        threads.push(thread);
    }
    for thread in threads {
        thread.join().expect("failed to join thread");
    }

    // decrement threads
    println!("starting {} decrement threads", num_threads);
    let mut threads = vec![];
    for thread_id in 0..num_threads {
        let thread = std::thread::Builder::new()
            .name(format!("dec{}", thread_id))
            .spawn(move || decrement(thread_id))
            .unwrap();
        threads.push(thread);
    }
    for thread in threads {
        thread.join().expect("failed to join thread");
    }
}


examples/profiler.rs
#![feature(thread_local)]

#[cfg(target_os = "linux")]
mod profiler {
    use std::io::Write;

    use anyhow::{Context, Result};
    use nanorand::{Rng, WyRand};
    use nix::libc;
    use nix::sched::*;
    use nix::unistd::*;

    const BASE_VAL: i64 = 1_000_000;
    const MAX_THREADS: i64 = 2;
    static mut CI_INTERVAL: i64 = 10_000_000;

    #[thread_local]
    #[allow(non_upper_case_globals)]
    static mut prev_ic: i64 = 0;

    #[thread_local]
    #[allow(non_upper_case_globals)]
    static mut prev_tsc: u64 = 0;

    #[thread_local]
    #[allow(non_upper_case_globals)]
    static mut buffer_ic: Vec<i64> = Vec::new();

    #[thread_local]
    #[allow(non_upper_case_globals)]
    static mut buffer_tsc: Vec<u64> = Vec::new();

    fn rand() -> i64 {
        let mut rng = WyRand::new();
        rng.generate_range(0..i64::MAX)
    }

    fn interrupt_handler(curr_ic: i64) {
        unsafe {
            let ic = curr_ic - prev_ic;
            if ic < 0 {
                panic!("IR count was negative: {}", ic);
            }

            let mut aux: u32 = 0;
            let curr_tsc = std::arch::x86_64::__rdtscp(&mut aux);
            let tsc = curr_tsc - prev_tsc;

            buffer_ic.push(ic);
            buffer_tsc.push(tsc);

            prev_ic = curr_ic;
            prev_tsc = curr_tsc;
        }
    }

    fn log_intervals() -> Result<()> {
        unsafe {
            compiler_interrupts::deregister();
        }

        let thread = std::thread::current();
        let thread_name = thread.name().context("failed to get thread name")?;

        let filename = format!("{}_intervals.txt", thread_name);

        let len: usize;
        let ordered_ic: Vec<i64>;
        let ordered_tsc: Vec<u64>;
        unsafe {
            len = buffer_tsc.len();

            // sort both buffer_ic and buffer_tsc using buffer_tsc as the key
            let mut indicies: Vec<usize> = (0..len).collect();
            indicies.sort_by_key(|&i| buffer_tsc[i]);
            ordered_ic = indicies.iter().map(|&i| buffer_ic[i]).collect();
            ordered_tsc = indicies.iter().map(|&i| buffer_tsc[i]).collect();
        }

        let file = std::fs::File::create(filename)?;
        let mut buf = std::io::BufWriter::new(file);
        writeln!(buf, "percentile, time-stamp counter, instruction count\n")?;
        for i in 0..len {
            let percentage = i as f64 / (len - 1) as f64;
            writeln!(
                buf,
                "{:.5}, {}, {}\n",
                percentage, ordered_tsc[i], ordered_ic[i]
            )?;
        }

        unsafe {
            if !buffer_tsc.is_empty() {
                let i = buffer_tsc.len() / 2;
                println!(
                    "thread: {} -> median interval: {} cycles",
                    thread_name, buffer_tsc[i]
                );
            }
        }

        Ok(())
    }

    fn pin_thread() -> Result<()> {
        // POSIX standard doesn't define `_SC_NPROCESSORS_CONF`
        let var = unsafe { std::mem::transmute::<i32, SysconfVar>(libc::_SC_NPROCESSORS_CONF) };
        let max_cpus = sysconf(var)?.context("failed to get number of cpus")?;
        let cpu = gettid().as_raw() % (max_cpus as i32 - 1);
        let mut cpu_set = CpuSet::new();
        cpu_set.set(cpu as usize)?;
        sched_setaffinity(Pid::this(), &cpu_set)?;

        Ok(())
    }

    fn increment() -> Result<()> {
        pin_thread()?;

        unsafe {
            compiler_interrupts::register(CI_INTERVAL, CI_INTERVAL, interrupt_handler);
        }

        let mut counter = 0;
        let iterations = BASE_VAL + (rand() % 10);
        for _ in 0..iterations {
            counter += rand() % 10;
        }

        log_intervals()?;

        println!(
            "thread: {} -> counter: {}",
            std::thread::current()
                .name()
                .context("failed to get thread name")?,
            counter
        );

        Ok(())
    }

    pub fn main() -> Result<()> {
        pin_thread()?;

        // check argument for CI interval
        let args = std::env::args().collect::<Vec<_>>();
        if args.len() == 2 {
            unsafe {
                // safety: CI_INTERVAL is only modified once before other threads read it
                CI_INTERVAL = args[1].parse().unwrap_or(CI_INTERVAL);
                println!("Using interrupt interval: {} IR", CI_INTERVAL)
            }
        } else {
            unsafe {
                println!("Using default interrupt interval: {} IR", CI_INTERVAL);
                println!("To change the interval: ./profiler <interval>");
            }
        }

        unsafe {
            compiler_interrupts::register(CI_INTERVAL, CI_INTERVAL, interrupt_handler);
        }

        println!("starting {} increment threads", MAX_THREADS);
        let mut threads = vec![];
        for thread_id in 0..MAX_THREADS {
            let thread = std::thread::Builder::new()
                .name(format!("inc{}", thread_id))
                .spawn(|| increment())
                .expect("failed to create thread");
            threads.push(thread);
        }
        for thread in threads {
            thread.join().expect("thread panicked")?;
        }

        log_intervals()?;

        println!("Achieved intervals (in cycles) per thread are exported to *_intervals.txt files");

        Ok(())
    }
}

#[cfg(target_os = "linux")]
fn main() -> anyhow::Result<()> {
    profiler::main()
}

#[cfg(not(target_os = "linux"))]
fn main() {
    println!("`profiler` example is available only on x86-64 Linux platforms");
}


src/lib.rs
//! [![crates.io](https://img.shields.io/crates/v/compiler-interrupts.svg)][crates.io]
//! [![docs.rs](https://docs.rs/compiler-interrupts/badge.svg)][docs.rs]
//! [![license](https://img.shields.io/crates/l/compiler-interrupts.svg)][license]
//!
//! `compiler-interrupts` provides Rust API for the Compiler Interrupts library.
//! Check out the Compiler Interrupts [main repository][compiler-interrupts] for more info.
//!
//! ## Requirements
//!
//! * [Rust 1.45.0][rust] or later is required.
//! Due to the usage of [`#[thread_local]`][thread_local] unstable feature,
//! this package currently requires nightly Rust.
//!
//! ## Getting started
//!
//! Add this to your `Cargo.toml`.
//!
//! ``` toml
//! [dependencies]
//! compiler-interrupts = "1.0"
//! ```
//!
//! Register the Compiler Interrupts handler in your program.
//!
//! ``` rust
//! #![feature(thread_local)]
//!
//! #[thread_local]
//! #[allow(non_upper_case_globals)]
//! static mut prev_ic: i64 = 0;
//!
//! fn interrupt_handler(ic: i64) {
//!     let interval;
//!     unsafe {
//!         // save the last interval
//!         interval = ic - prev_ic;
//!
//!         // update the instruction count
//!         prev_ic = ic;
//!     }
//!     if interval < 0 {
//!         panic!("IR interval was negative")
//!     }
//!     println!(
//!         "CI @ {}: last interval = {} IR",
//!         std::thread::current().name().expect("invalid thread name"),
//!         interval
//!     );
//! }
//!
//! fn main() {
//!     // register the CI handler for 1000 IR and cycles interval
//!     unsafe {
//!         compiler_interrupts::register(1000, 1000, interrupt_handler);
//!     }
//!
//!     // do something compute-intensive
//!     for _ in 0..100 {}
//!
//!     println!("i can add an unsafe block, right?")
//! }
//! ```
//!
//! If you have [`cargo-compiler-interrupts`][cargo-compiler-interrupts] installed,
//! you can now run `cargo build-ci` to start the compilation and integration.
//! Check out the **[documentation][docs.rs]** for more info about the API.
//!
//! ## Contribution
//!
//! All issue reports, feature requests, pull requests and GitHub stars are welcomed
//! and much appreciated.
//!
//! ## Author
//!
//! Quan Tran ([@quanshousio][quanshousio])
//!
//! ## Acknowledgements
//!
//! * My advisor [Jakob Eriksson][jakob] for the enormous support for this project.
//! * [Nilanjana Basu][nilanjana] for implementing the Compiler Interrupts.
//!
//! ## License
//!
//! `compiler-interrupts` is available under the MIT license.
//! See the [LICENSE][license] file for more info.
//!
//! [crates.io]: https://crates.io/crates/compiler-interrupts
//! [docs.rs]: https://docs.rs/compiler-interrupts
//! [license]: https://github.com/bitslab/compiler-interrupts-rs/blob/main/LICENSE
//! [compiler-interrupts]: https://github.com/bitslab/CompilerInterrupts
//! [rust]: https://www.rust-lang.org/tools/install
//! [thread_local]: https://github.com/rust-lang/rust/issues/29594
//! [cargo-compiler-interrupts]: https://github.com/bitslab/cargo-compiler-interrupts
//! [quanshousio]: https://quanshousio.com
//! [jakob]: https://www.linkedin.com/in/erikssonjakob
//! [nilanjana]: https://www.linkedin.com/in/nilanjana-basu-99027959

#![feature(thread_local)]

/// Default large interval
const LARGE_INTERVAL: i64 = 100000;

/// Default small interval
const SMALL_INTERVAL: i64 = 10000;

/// Interrupt function for the framework.
#[no_mangle]
#[thread_local]
static mut intvActionHook: fn(i64) = dummy;

/// Store the interrupt handler from [`register`].
#[allow(non_upper_case_globals)]
#[thread_local]
static mut int_handler: fn(i64) = dummy;

/// Store the enable hook from [`register_enable_hook`].
#[allow(non_upper_case_globals)]
#[thread_local]
static mut enableHook: Option<fn()> = None;

/// Store the disable hook from [`register_disable_hook`].
#[allow(non_upper_case_globals)]
#[thread_local]
static mut disableHook: Option<fn()> = None;

/// IR interrupt interval for the framework.
#[no_mangle]
#[thread_local]
static mut ci_ir_interval: i64 = LARGE_INTERVAL;

/// IR interrupt reset interval when target target cycles is not exceeded
/// for the framework.
#[no_mangle]
#[thread_local]
static mut ci_reset_ir_interval: i64 = LARGE_INTERVAL / 2;

/// Cycles interrupt interval for the framework.
#[no_mangle]
#[thread_local]
static mut ci_cycles_interval: i64 = SMALL_INTERVAL;

/// Cycles interrupt threshold to fire the interrupt or reset the IR counter
/// for the framework.
#[no_mangle]
#[thread_local]
static mut ci_cycles_threshold: i64 = (0.9 * LARGE_INTERVAL as f64) as i64;

/// Thread-local local counter for the framework.
#[no_mangle]
#[thread_local]
static mut LocalLC: i32 = 0;

/// Thread-local disable counter for the framework.
#[no_mangle]
#[thread_local]
static mut lc_disabled_count: i32 = 0;

/// Thread-local next interval for the framework.
#[no_mangle]
#[thread_local]
static mut NextInterval: i32 = 0;

/// A dummy function.
fn dummy(_: i64) {}

/// Assigns the interrupt function to itself and calls the handler from [`register`].
fn interrupt_handler(ic: i64) {
    unsafe {
        intvActionHook = dummy;
        int_handler(ic);
        intvActionHook = interrupt_handler
    }
}

/// Registers a handler for Compiler Interrupts.
///
/// This function takes a IR interval, cycles interval, and
/// function pointer to the Compiler Interrupts handler.
/// The handler receives an approximation of the number of IR instructions
/// since the last interrupt as the argument.
///
/// # Note
///
/// This function is thread-specific, which means it only registers
/// on the thread they called on.
///
/// This function should not be called multiple times.
/// Consecutive calls will override the previous intervals and handler.
///
/// # Safety
///
/// This function mutates a thread-local static variable which uses for the interrupt handler.
/// Thread unsafety will not be introduced. However, calling the handler outside Rust would
/// probably violate Rust's safe memory model; hence the function is considered unsafe.
///
/// # Examples
///
/// ```
/// fn interrupt_handler(ic: i64) {
///     println!("Compiler interrupt called with instruction count: {}", ic);
/// }
///
/// unsafe {
///     compiler_interrupts::register(10000, 10000, interrupt_handler);
/// }
/// ```
pub unsafe fn register(ir_interval: i64, cycles_interval: i64, handler: fn(i64)) {
    LocalLC += ci_ir_interval as i32;
    ci_ir_interval = ir_interval;
    ci_reset_ir_interval = ir_interval / 2;
    ci_cycles_interval = cycles_interval;
    ci_cycles_threshold = (0.9 * cycles_interval as f64) as i64;
    int_handler = handler;
    intvActionHook = interrupt_handler;
}

/// De-registers the handler for Compiler Interrupts.
///
/// This function removes the given interrupts handler from [`register`].
///
/// # Note
///
/// This function is thread-specific, which means it only de-registers
/// on the thread they called on.
///
/// This function should not be called multiple times.
/// Consecutive calls will do nothing as the handler has already been de-registered.
///
/// # Safety
///
/// This function mutates a thread-local static variable which uses for the interrupt handler.
/// Thread unsafety will not be introduced. Rust considers mutating static variable unsafe.
pub unsafe fn deregister() {
    ci_ir_interval = LARGE_INTERVAL;
    ci_reset_ir_interval = LARGE_INTERVAL / 2;
    ci_cycles_interval = LARGE_INTERVAL;
    ci_cycles_threshold = (0.9 * LARGE_INTERVAL as f64) as i64;
    int_handler = dummy;
    intvActionHook = dummy;
}

/// Enables Compiler Interrupts.
///
/// This function enables Compiler Interrupts if it is currently disabled.
/// Compiler Interrupts are enabled by default.
///
/// # Note
///
/// This function is thread-specific, which means it only enables
/// on the thread they called on.
///
/// This function can be called multiple times.
/// Number of [`enable`] calls must be the same as the number of previous [`disable`] calls
/// to re-enable the interrupts.
///
/// # Safety
///
/// This function mutates a thread-local static variable which uses for the counter.
/// Thread unsafety will not be introduced. Rust considers mutating static variable unsafe.
///
/// # Examples
///
/// ```
/// unsafe {
///     // disables the interrupts
///     compiler_interrupts::disable();
/// }
///
/// for _ in 0..42 {
///     println!("interrupts have been disabled");
/// }
///
/// unsafe {
///     // re-enables the interrupts
///     compiler_interrupts::enable();
/// }
/// ```
pub unsafe fn enable() {
    if lc_disabled_count > 0 {
        lc_disabled_count -= 1;
    }
    if let Some(hook) = enableHook {
        hook();
    }
    if lc_disabled_count == 0 {
        intvActionHook = interrupt_handler;
    }
}

/// Disables Compiler Interrupts.
///
/// This function disables Compiler Interrupts if it is currently enabled.
///
/// # Note
///
/// This function is thread-specific, which means it only disables
/// on the thread they called on.
///
/// This function can be called multiple times.
/// Consecutive calls will do nothing as the interrupts are already disabled.
///
/// # Safety
///
/// This function mutates a thread-local static variable which uses for the counter.
/// Thread unsafety will not be introduced. Rust considers mutating static variable unsafe.
///
/// # Examples
///
/// ```
/// unsafe {
///     // disables the interrupts
///     compiler_interrupts::disable();
/// }
///
/// for _ in 0..42 {
///     println!("interrupts have been disabled");
/// }
///
/// unsafe {
///     // re-enables the interrupts
///     compiler_interrupts::enable();
/// }
/// ```
pub unsafe fn disable() {
    intvActionHook = dummy;
    lc_disabled_count += 1;
    if let Some(hook) = disableHook {
        hook();
    }
}

/// Registers a hook when enabling Compiler Interrupts.
///
/// This function takes a function pointer to be called after enabling Compiler Interrupts.
/// Compiler Interrupts can be enabled by calling [`enable`].
///
/// # Note
///
/// This function is thread-specific, which means it only registers
/// on the thread they called on.
///
/// This function should not be called multiple times.
/// Consecutive calls will override the previous hook.
///
/// # Safety
///
/// This function mutates a thread-local static variable which uses for the hook.
/// Thread unsafety will not be introduced. Rust considers mutating static variable unsafe.
pub unsafe fn register_enable_hook(hook: fn()) {
    enableHook = Some(hook)
}

/// De-registers the hook when enabling Compiler Interrupts.
///
/// This function removes the given hook from [`register_enable_hook`].
///
/// # Note
///
/// This function is thread-specific, which means it only de-registers
/// on the thread they called on.
///
/// This function should not be called multiple times.
/// Consecutive calls will do nothing as the hook has already been removed.
///
/// # Safety
///
/// This function mutates a thread-local static variable which uses for the hook.
/// Thread unsafety will not be introduced. Rust considers mutating static variable unsafe.
pub unsafe fn deregister_enable_hook() {
    enableHook = None
}

/// Registers a hook when disabling Compiler Interrupts.
///
/// This function takes a function pointer to be called before disabling Compiler Interrupts.
/// Compiler Interrupts can be temporarily disabled by calling [`disable`].
///
/// # Note
///
/// This function is thread-specific, which means it only registers
/// on the thread they called on.
///
/// This function should not be called multiple times.
/// Consecutive calls will override the previous hook.
///
/// # Safety
///
/// This function mutates a thread-local static variable which uses for the hook.
/// Thread unsafety will not be introduced. Rust considers mutating static variable unsafe.
pub unsafe fn register_disable_hook(hook: fn()) {
    disableHook = Some(hook)
}

/// De-registers the hook when disabling Compiler Interrupts.
///
/// This function removes the given hook from [`register_disable_hook`].
///
/// # Note
///
/// This function is thread-specific, which means it only de-registers
/// on the thread they called on.
///
/// This function should not be called multiple times.
/// Consecutive calls will do nothing as the hook has already been removed.
///
/// # Safety
///
/// This function mutates a thread-local static variable which uses for the hook.
/// Thread unsafety will not be introduced. Rust considers mutating static variable unsafe.
pub unsafe fn deregister_disable_hook() {
    disableHook = None
}

/// Enables the probe instrumentation.
///
/// # Note
///
/// This function is thread-specific, which means it only enables
/// on the thread they called on.
///
/// # Safety
///
/// This function is called outside the normal Rust program.
#[no_mangle]
pub unsafe fn instr_enable() {}

/// Disables the probe instrumentation.
///
/// # Note
///
/// This function is thread-specific, which means it only disables
/// on the thread they called on.
///
/// # Safety
///
/// This function is called outside the normal Rust program.
#[no_mangle]
pub unsafe fn instr_disable() {}
